drivers/gpio/gpio_sim.c

```diff
@@ -0,0 +1,563 @@
+/*
+ * Copyright (c) 2020 Friedt Professional Engineering Services, Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT zephyr_sim_gpio
+
+#include <device.h>
+#include <drivers/gpio.h>
+#include <drivers/gpio/gpio_sim.h>
+#include <errno.h>
+#include <stdio.h>
+#include <zephyr.h>
+
+#include "gpio_utils.h"
+
+#define LOG_LEVEL CONFIG_GPIO_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(gpio_sim);
+
+#define interrupt_bitmask                                                      \
+	(0 | GPIO_INT_DISABLE | GPIO_INT_ENABLE | GPIO_INT_LEVELS_LOGICAL |    \
+	 GPIO_INT_EDGE | GPIO_INT_LOW_0 | GPIO_INT_HIGH_1)
+
+/*
+ * GPIO config / data
+ */
+
+struct gpio_sim_config {
+	struct gpio_driver_config common;
+	struct k_mutex mu;
+	sys_slist_t callbacks;
+	uint32_t flags[CONFIG_GPIO_SIM_NPINS_PER_PORT];
+	gpio_port_value_t vals;
+	gpio_port_pins_t interrupts;
+};
+
+struct gpio_sim_data {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+	const struct gpio_sim_config *config;
+};
+
+static struct gpio_sim_config gpio_sim_config[CONFIG_GPIO_SIM_NPORTS];
+static struct gpio_sim_data gpio_sim_data[CONFIG_GPIO_SIM_NPORTS];
+
+/*
+ * Util
+ */
+static gpio_port_pins_t
+gpio_sim_get_mask_of_pins_with_flags(struct gpio_sim_config *config,
+				     uint32_t flags)
+{
+	size_t i;
+	gpio_port_pins_t mask;
+
+	for (mask = 0, i = 0; i < CONFIG_GPIO_SIM_NPINS_PER_PORT; ++i) {
+		if (config->flags[i] & flags) {
+			mask |= (1 << i);
+		}
+	}
+
+	return mask;
+}
+
+static gpio_port_pins_t gpio_sim_input_mask(struct gpio_sim_config *config)
+{
+	return gpio_sim_get_mask_of_pins_with_flags(config, GPIO_INPUT);
+}
+
+static gpio_port_pins_t gpio_sim_output_mask(struct gpio_sim_config *config)
+{
+	return gpio_sim_get_mask_of_pins_with_flags(config, GPIO_OUTPUT);
+}
+
+/*
+ * GPIO backend API (for setting input pin values)
+ */
+
+static void gpio_sim_pend_interrupt(struct device *port, gpio_port_pins_t mask,
+				    gpio_port_value_t prev_values,
+				    gpio_port_value_t values)
+{
+	size_t i;
+	struct gpio_sim_data *drv_data =
+		(struct gpio_sim_data *)port->driver_data;
+	struct gpio_sim_config *config =
+		(struct gpio_sim_config *)drv_data->config;
+	gpio_port_pins_t interrupts = 0;
+
+	for (i = 0; mask && i < CONFIG_GPIO_SIM_NPINS_PER_PORT;
+	     ++i, mask >>= 1, prev_values >>= 1, values >>= 1) {
+		if (!(mask & 1)) {
+			continue;
+		}
+
+		bool prev_bit = prev_values & 1;
+		bool bit = values & 1;
+
+		switch (config->flags[i] & interrupt_bitmask) {
+		case GPIO_INT_EDGE_RISING:
+			if (!prev_bit && bit) {
+				interrupts |= (1 << i);
+			}
+			break;
+
```

Comment:
Are we supposed to have blank lines here?

Reply:
None of the linters seem to have complained. It was formatted with clang-format using Zephyr's .clang-format file.

Reply:
Removed the blank lines